src/partition/ReceivedPartition.cpp

```diff
@@ -464,145 +466,330 @@ void ReceivedPartition::createOwnerInformation()
-    slaveOwnerVecs[0].resize(_mesh->vertices().size());
-    slaveGlobalIDs[0].resize(_mesh->vertices().size());
-    slaveTags[0].resize(_mesh->vertices().size());
-    for (size_t i = 0; i < _mesh->vertices().size(); i++) {
-      slaveGlobalIDs[0][i] = _mesh->vertices()[i].getGlobalIndex();
-      if (_mesh->vertices()[i].isTagged()) {
-        masterAtInterface = true;
-        slaveTags[0][i]   = 1;
-      } else {
-        slaveTags[0][i] = 0;
-      }
+    // #4: Exchange number of already owned vertices with the neighbors
+    std::map<int, int> neighborRanksVertexCount; // for load balancing
+    int receivedCounter = 0;
+    for (auto & neighborRank: localConnectedBBMap)
+    {
+      utils::MasterSlave::_communication->aSend(counter, neighborRank.first);
+    }
+    
+    for (auto & neighborRank: localConnectedBBMap)
+    {
+      utils::MasterSlave::_communication->receive(receivedCounter, neighborRank.first);
+      neighborRanksVertexCount.emplace(neighborRank.first, receivedCounter);
     }
-    PRECICE_DEBUG("My tags: " << slaveTags[0]);
 
-    // receive slave data
-    int ranksAtInterface = 0;
-    if (masterAtInterface)
-      ranksAtInterface++;
+    // #5: Exchange list of shared vertices with the neighbor
+    for (auto & receivingRank: sharedVerticesSendMap)
+    {
+      int sendSize = receivingRank.second.size();
+      utils::MasterSlave::_communication->aSend(sendSize, receivingRank.first);
+      if (sendSize != 0)
+      {
+        utils::MasterSlave::_communication->aSend(receivingRank.second, receivingRank.first);
+      }
+    }
 
-    for (int rank = 1; rank < utils::MasterSlave::getSize(); rank++) {
-      int localNumberOfVertices = -1;
-      utils::MasterSlave::_communication->receive(localNumberOfVertices, rank);
-      PRECICE_DEBUG("Rank " << rank << " has " << localNumberOfVertices << " vertices.");
-      slaveOwnerVecs[rank].resize(localNumberOfVertices, 0);
-
-      if (localNumberOfVertices != 0) {
-        PRECICE_DEBUG("Receive tags from slave rank " << rank);
-        utils::MasterSlave::_communication->receive(slaveTags[rank], rank);
-        utils::MasterSlave::_communication->receive(slaveGlobalIDs[rank], rank);
-        PRECICE_DEBUG("Rank " << rank << " has tags " << slaveTags[rank]);
-        PRECICE_DEBUG("Rank " << rank << " has global IDs " << slaveGlobalIDs[rank]);
-        bool atInterface = false;
-        utils::MasterSlave::_communication->receive(atInterface, rank);
-        if (atInterface)
-          ranksAtInterface++;
+    for (auto & neighborRank: sharedVerticesSendMap)
+    {
+      int receiveSize = 0;
+      utils::MasterSlave::_communication->receive(receiveSize, neighborRank.first);
+      if (receiveSize != 0)
+      {
+        std::vector<double> receivedSharedVertices;
+        utils::MasterSlave::_communication->receive(receivedSharedVertices, neighborRank.first);
+        sharedVerticesReceiveMap.insert(std::make_pair(neighborRank.first, receivedSharedVertices));
```

Comment:
I still do not completely understand why aSend/receive can result in a deadlock but send/aReceive won't. 

The problem with aReceive/send is, that we have to separate the communication of number of vertices and vertices list. This is because, we first need to check if there is any vertex to communicate. If we do them together, the list size would always be 0, since it takes time to aReceive gets it,  and therefore, it will not enter to the if where it receives the actual list. 

If we separate the size and list communication, we would need to define another map to save the message size for each rank separately. This would be ugly in my opinion, and I would avoid it unless we make sure that the deadlock threat is real ;)  

Regarding the check after aSend, I think it is not necessary, since the aSend is followed by a receive. This is an implicit barrier, right?

Reply:
aSend uses `MPI_Isend` under the hood. There is no guarantee that this function is asynchronous. For large messages (such as the list of shared vertices at some point?) it might block. More information [here](https://www.mcs.anl.gov/research/projects/mpi/sendmode.html) for example.
Whenever we use an `aSend` or an `aReceive`, we should always check the request. Otherwise this is prone to fail if somebody else further develops this code part. You cannot be sure that the implicit barrier will always be there. Also, MPI does not give any guarantee that this implicit barrier is enough.
If we don't need the request, we could also just use a normal `send`.